src/Kupo/App/ChainSync/Hydra.hs

```diff
@@ -0,0 +1,113 @@
+--  This Source Code Form is subject to the terms of the Mozilla Public
+--  License, v. 2.0. If a copy of the MPL was not distributed with this
+--  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+module Kupo.App.ChainSync.Hydra
+    ( connect
+    , runChainSyncClient
+    , newTransactionStore
+    , TransactionStore (..)
+    , TransactionStoreException (..)
+    ) where
+
+import Kupo.Prelude
+
+import Control.Exception.Safe
+    ( MonadThrow
+    , throwM
+    )
+import qualified Data.Map as Map
+import Kupo.App.Mailbox
+    ( Mailbox
+    , putHighFrequencyMessage
+    )
+import Kupo.Control.MonadSTM
+    ( MonadSTM (..)
+    )
+import Kupo.Data.Cardano
+    ( Point
+    , Tip
+    , TransactionId
+    )
+import Kupo.Data.ChainSync
+    ( IntersectionNotFoundException (..)
+    )
+import Kupo.Data.Hydra
+    ( HydraMessage (..)
+    , Snapshot (..)
+    , decodeHydraMessage
+    , mkHydraBlock
+    )
+import Kupo.Data.PartialBlock
+    ( PartialBlock
+    , PartialTransaction (..)
+    )
+import qualified Network.WebSockets as WS
+import qualified Network.WebSockets.Json as WS
+
+runChainSyncClient
+    :: forall m.
+        ( MonadIO m
+        , MonadSTM m
+        , MonadThrow m, MonadThrow (STM m))
+    => Mailbox m (Tip, PartialBlock) (Tip, Point)
+    -> m () -- An action to run before the main loop starts.
+    -> [Point]
+    -> WS.Connection
+    -> m IntersectionNotFoundException
+runChainSyncClient mailbox beforeMainLoop _pts ws = do
+    beforeMainLoop
+    TransactionStore{pushTx, popTxById} <- newTransactionStore
+    forever $ do
+        WS.receiveJson ws decodeHydraMessage >>= \case
+            HeadIsOpen{genesisTxs} ->
+                atomically (putHighFrequencyMessage mailbox (mkHydraBlock 0 genesisTxs))
+            TxValid{tx} ->
+                pushTx tx
```

Comment:
I suppose that this will work for lack of better options, but this will cause Kupo's memory to grow possibly unbounded. 

Transactions are 'popped' on snapshot confirmed, which means that the memory should likely just grow and shrink on every snapshot, but that also means that any non-confirmed transaction will sit in the memory forever.

Reply:
That's true. It's not ideal, but necessary given the current API. We should, however, record kupo's use case for a need of a more "complete" `SnapshotConfirmed` message.. there is an ADR about making the Hydra API more resource based: https://github.com/input-output-hk/hydra/pull/1028.